### PR TITLE
#24 - Use script tag for prompts and admin path

### LINKED ISSUE #24: Should we be assigning JS variables to the window global?
https://github.com/wagtail/wagtail-ai/blob/04bc8c3d4a88a20dd6f1d6642ddbed758035d5cd/src/wagtail_ai/wagtail_hooks.py#L84-L97

I think we could use Django's json_script instead (https://github.com/django/django/blob/61c305f298da1b4079a80721c861d0663dc8717e/django/utils/html.py#L85-L102). There's no need to write JS at all to pass those values to the JS and it might be problematic with CSP as well.

### PR DESCRIPTION
**Pull request by ianmeigh**
The Admin path used by the editor magic wand and the prompts are no longer assigned as JavaScript variables in the window global and get added to the DOM in a script tag (id = wagtail-ai-config).

Closes #24.

### CODE REVIEW

File: src/wagtail_ai/static_src/custom.d.ts [hunk 1]
```diff
@@ -1,6 +1,11 @@
 /* eslint-disable no-unused-vars */
 export {};
 
+export interface WagtailAiConfiguration {
+  wagtailAiPrompts: Array<Prompt>;
```
Comment (tm-kn) [inline src/wagtail_ai/static_src/custom.d.ts:5]: You can just use the `Prompt[]` syntax.

File: src/wagtail_ai/static_src/custom.d.ts [hunk 2]
```diff
@@ -1,6 +1,11 @@
...
+  wagtailAiProcessUrl: string;
```
Comment (tm-kn) [inline src/wagtail_ai/static_src/custom.d.ts:6]: **Suggestion:** there's no need to prefix all the settings with `WagtailAi` given the interface already has it in the name.

File: src/wagtail_ai/static_src/utils.tsx [hunk 3]
```diff
@@ -19,7 +35,9 @@ const fetchAIResponse = async (
   formData.append('text', text);
   formData.append('prompt', prompt.uuid);
   try {
-    const res = await fetch(window.WAGTAIL_AI_PROCESS_URL, {
+    const aiProcessUrl = getAIConfiguration()?.wagtailAiProcessUrl;
```
Comment (tm-kn) [inline src/wagtail_ai/static_src/utils.tsx:38]: **Suggestion:** There's no need for `?` given GetAIConfiguration always returns a value thanks to its type.

File: src/wagtail_ai/wagtail_hooks.py [hunk 4]
```diff
@@ -83,18 +83,17 @@ def get_prompts():
...
-        f"""
-        <script>
-            window.WAGTAIL_AI_PROCESS_URL = "{process_url}";
-            window.WAGTAIL_AI_PROMPTS = {prompt_json};
-        </script>
-        """
+    wagtail_ai_config = json_script(
+        {"wagtail_ai_prompts": prompt_json, "wagtail_ai_process_url": process_url},
+        "wagtail-ai-config",
+        DjangoJSONEncoder,
```
Comment (tm-kn) [inline src/wagtail_ai/wagtail_hooks.py:92]: I think this should be a keyword argument.
```suggestion
        encoder=DjangoJSONEncoder,
```

File: src/wagtail_ai/static_src/AIControl.tsx [hunk 5]
```diff
@@ -70,6 +77,7 @@ function ToolbarDropdown({
 }
 
 function AIControl({ getEditorState, onChange }: ControlComponentProps) {
+  const aiPrompts = getAIConfiguration()?.wagtailAiPrompts as Array<Prompt>;
```
Comment (tm-kn) [inline src/wagtail_ai/static_src/AIControl.tsx:80]: **Issue:** Why do you need to use type casting (`as`) if wagtailAiPrompts is always an array? Don't use casting unless you have to.

File: src/wagtail_ai/wagtail_hooks.py (hunk 4 above)
Comment (ianmeigh) [reply src/wagtail_ai/wagtail_hooks.py:92]: Changing this to a keyword argument doesn't fix the type issue that's been raised by Pyright.

I think it's related to [django-types](https://github.com/sbdchd/django-types/blob/d44899f411df6036610b81cc06cddccd8f1e17cd/django-stubs/utils/html.pyi#L17) only listing two arguments for json_script, whereas it should accept 3 ([django repo link](https://github.com/django/django/blob/main/django/utils/html.py#L85)).

File: src/wagtail_ai/static_src/utils.tsx [hunk 6]
```diff
@@ -5,11 +5,27 @@ import {
...
 
+export const getAIConfiguration = (): WagtailAiConfiguration => {
+  const configurationElement =
+    document.querySelector<HTMLScriptElement>('#wagtail-ai-config');
+  if (!configurationElement || !configurationElement.textContent) {
+    throw Error('No wagtail-ai configuration found.');
+  }
+
+  try {
+    return camelizeKeys(
```
Comment (tomusher) [inline src/wagtail_ai/static_src/utils.tsx:21]: **Suggestion:** The addition of a dependency just for this doesn't seem worth it. If we want camel case in JS, we could provide an already camel-case'd dict from Python?

File: src/wagtail_ai/static_src/custom.d.ts (hunk 1 above)
Comment (ianmeigh) [reply src/wagtail_ai/static_src/custom.d.ts]: Resoved in def70f2e2fa386e2fcaf08371da29c0e83b0e82e.

File: src/wagtail_ai/static_src/custom.d.ts (hunk 2 above)
Comment (ianmeigh) [reply src/wagtail_ai/static_src/custom.d.ts]: Renamed in f4890e2952f17de4f10326eacf7bef560f7d0a8c.

File: src/wagtail_ai/static_src/utils.tsx (hunk 3 above)
Comment (ianmeigh) [reply src/wagtail_ai/static_src/utils.tsx]: Removed in https://github.com/wagtail/wagtail-ai/commit/17298857b64c6fa1c72f307764522751b3b81fac.

File: src/wagtail_ai/wagtail_hooks.py (hunk 4 above)
Comment (tomusher) [reply src/wagtail_ai/wagtail_hooks.py:92]: Good catch. As the default encoder for `json_script` is `DjangoJSONEncoder` we can probably just drop that argument here?

Comment (ianmeigh) [reply src/wagtail_ai/wagtail_hooks.py:92]: Thanks Tom, I missed that while I was trying to find out why the type check didn't work.

File: src/wagtail_ai/static_src/utils.tsx (hunk 6 above)
Comment (ianmeigh) [reply src/wagtail_ai/static_src/utils.tsx]: Refactored in 73506050e4cd6b27dc7b541ed96f0c5747425842 and removed dependency.

File: src/wagtail_ai/wagtail_hooks.py (hunk 4 above)
Comment (ianmeigh) [reply src/wagtail_ai/wagtail_hooks.py]: Resolved in 81985f16ee6886d79a95732f928e42a9bb1a6bbc.

File: src/wagtail_ai/static_src/utils.tsx [hunk 7]
```diff
@@ -5,11 +5,27 @@ import {
...
+      JSON.parse(configurationElement.textContent),
+    ) as WagtailAiConfiguration;
+  } catch (err) {
+    throw Error(err);
```
[the remark this answers is not in the document]
Comment (ianmeigh) [reply src/wagtail_ai/static_src/utils.tsx]: I have updated the error message in e13e0290ab1cb21f424315ed1d6209c7f1bdc3af.

**Comment by tomusher** [comment]
Thanks @ianmeigh!

**Pull request merged**